Update AC Workflow and generator util classes

gigxz opened the pull request:
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch
- use a merge-commit or rebase strategy for PRs targeting the stable branch

## Description
Issue: https://github.com/open-path/Green-River/issues/8007

* Update Housing Workflow. Move form definitions out of rake task and into `json` files.
* Add Admin Assign Workflow for direct referrals. This workflow has lots of shared logic with the housing workflow, so the commonalities are pulled into `build_provider_outcome_denial_review_loop`
* Refactor workflow-builder utility class from rake file into own class
* Refactor methods for building AC workflows into dedicated class

## Type of change
Configuration, new feature

## How to test
```
CLIENT=x rails driver:hmis:seed_definitions
rails driver:hmis:ce_define_workflows
```

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] If adding a new endpoint / exposing data in a new way, I have:
- [x] Any major architectural changes are supported by an approved ADR (Architectural Decision Record) 
- [x] I have updated the documentation (or not applicable)
- [ ] I have added spec tests (or not applicable)
- [x] I have provided testing instructions in this PR or the related issue (or not applicable)

gigxz committed:
wip

gigxz committed:
wip

gigxz committed:
Updated forms

gigxz committed:
wip

gigxz committed:
wip

gigxz committed:
Add admin assign workflow

gigxz committed:
refactor wip

gigxz committed:
refactor

gigxz committed:
Self review

gigxz committed:
Merge branch 'release-174' into gig/workflow-updates-ac

gigxz committed:
Self review

gigxz commented:
retarget to 175

gigxz commented:
Here are the hand-written simplified diagrams. These show how everything beyond "Provider Outcome" is the same between the 2 workflows.

**Housing Workflow**
yellow = "matching in progress"
blue = "assigned"
pink = "denial pending"
```mermaid
flowchart TD
start_referral(("Start Referral"))
start_referral--> initial_review
initial_review("Initial Review"):::yellow_node
initial_review--> referral_declined
initial_review--> client_engagement_start_date
referral_declined(("Referral Declined"))
client_engagement_start_date("Initial Client Engagement Start Date"):::yellow_node
client_engagement_start_date--> client_engagement
client_engagement("Client Engagement"):::yellow_node
client_engagement-->client_offer_outcome
client_offer_outcome("Client Offer Outcome"):::yellow_node
client_offer_outcome--> referral_declined
client_offer_outcome--> provider_outcome
provider_outcome("Provider Outcome"):::blue_node
provider_outcome--> denial_review
provider_outcome--Client Enrolled-->confirm_success
denial_review("Denial Review"):::pink_node
denial_review--> referral_declined
denial_review--> provider_outcome_2
confirm_success--> referral_declined
provider_outcome_2("Provider Outcome (2)"):::blue_node
provider_outcome_2--> denial_review_2
provider_outcome_2--Client Enrolled-->confirm_success
denial_review_2("Denial Review (2)"):::pink_node
denial_review_3("Denial Review (3)"):::pink_node
denial_review_3-->referral_declined
denial_review_2-->provider_outcome_3
denial_review_2-->referral_declined
provider_outcome_3("Provider Outcome (3)"):::blue_node
provider_outcome_3-->denial_review_3
provider_outcome_3--Client Enrolled-->confirm_success
client_engagement-->referral_declined
confirm_success("Confirm Success")
confirm_success--> referral_accepted
referral_accepted(("Referral Accepted"))

classDef pink_node fill:#ffe6e6,stroke:#ffcccc,stroke-width:2px;
classDef yellow_node fill:#ffffcc,stroke:#ffeb99,stroke-width:2px;
classDef blue_node fill:#e6f7ff,stroke:#99d6ff,stroke-width:2px;

```


**Direct Referral Workflow**
```mermaid
flowchart TD
start_referral(("Start Referral"))
start_referral--> initial_review
referral_declined(("Referral Declined"))
initial_review("Admin Assign"):::yellow_node
initial_review-->provider_outcome
provider_outcome("Provider Outcome"):::blue_node
provider_outcome--> denial_review
provider_outcome--Client Enrolled-->confirm_success
denial_review("Denial Review"):::pink_node
denial_review--> referral_declined
denial_review--> provider_outcome_2
confirm_success--> referral_declined
provider_outcome_2("Provider Outcome (2)"):::blue_node
provider_outcome_2--> denial_review_2
provider_outcome_2--Client Enrolled-->confirm_success
denial_review_2("Denial Review (2)"):::pink_node
denial_review_3("Denial Review (3)"):::pink_node
denial_review_3-->referral_declined
denial_review_2-->provider_outcome_3
denial_review_2-->referral_declined
provider_outcome_3("Provider Outcome (3)"):::blue_node
provider_outcome_3-->denial_review_3
provider_outcome_3--Client Enrolled-->confirm_success
confirm_success("Confirm Success")
confirm_success--> referral_accepted
referral_accepted(("Referral Accepted"))

classDef pink_node fill:#ffe6e6,stroke:#ffcccc,stroke-width:2px;
classDef yellow_node fill:#ffffcc,stroke:#ffeb99,stroke-width:2px;
classDef blue_node fill:#e6f7ff,stroke:#99d6ff,stroke-width:2px;

```

gigxz commented on drivers/hmis/lib/ce_workflows/builder.rb:
```diff
@@ -0,0 +1,123 @@
+# Not intended for use in production application logic.
+#
+module CeWorkflows
+  class Builder
```
These methods are unchanged, just copy-pasted from `drivers/hmis/lib/tasks/ce_define_workflows.rake`

gigxz replied:
These all need better names. We have `HmisUtil::CeBuilder` which is a dev utility  with a couple methods, maybe we should combine them.

gigxz commented on drivers/hmis/lib/ce_workflows/ac_builder.rb:
```diff
@@ -0,0 +1,381 @@
+      initial_client_engagement: 'housing_workflow_initial_client_engagement',
+      client_engagement: 'housing_workflow_client_engagement',
+      client_offer_outcome: 'housing_workflow_client_offer_outcome',
+      provider_outcome_1: 'housing_workflow_provider_outcome_1',
+      provider_outcome_2: 'housing_workflow_provider_outcome_2',
+      provider_outcome_3: 'housing_workflow_provider_outcome_3',
```
decided to do separate forms for each iteration of the provider outcome task, even though forms are identical. we could probably use the same form if we want to. i'm unclear on benefits/pitfalls.

martha replied:
I think I'd lean towards using the same form, because it seems more likely to me that in the future the forms will need to change in parallel with each other, rather than needing to diverge from one another. But I think it's fine whatever you decide to do.

gigxz replied:
I think I'm a bit worried about the reporting implication of re-using the same CustomDataElementDefinition (ie `housing_workflow_provider_outcome_note`) across multiple steps. (If we used the same form for all 3 steps, then all 3 steps would be collecting CDEs using the same custom data element definition). That might be fine––each CDE will be "owned" by a different node instance–– but we'd need to figure out how to solve that for reporting. For example if we want to report out on all the CDEs that were collected on a Referral. It feels like it will be more straightforward to use unique CDEDs (`housing_workflow_provider_outcome_1_note`, `housing_workflow_provider_outcome_2_note`) to more easily distinguish where they were collected.

I definitely feel like we _should_ be able to support either way, but this felt safer to me.

gigxz replied:
added comments about this

gigxz commented on drivers/hmis/lib/ce_workflows/ac_builder.rb:
```diff
@@ -0,0 +1,381 @@
+      provider_outcome_3: 'housing_workflow_provider_outcome_3',
+      denial_review_1: 'housing_workflow_denial_review_1',
+      denial_review_2: 'housing_workflow_denial_review_2',
+      denial_review_3: 'housing_workflow_denial_review_3',
```
The third denial review is different from the others because it doesn't let you "send back". Otherwise, they are the same.

martha replied:
Worth adding as a comment in the code?

gigxz committed:
Merge branch 'release-174' into gig/workflow-updates-ac

gigxz committed:
Merge branch 'release-175' into gig/workflow-updates-ac

gigxz committed:
Merge branch 'release-175' into gig/workflow-updates-ac

gigxz committed:
Clean up

martha commented on drivers/hmis/lib/ce_workflows/ac_builder.rb:
```diff
@@ -0,0 +1,381 @@
+      raise "Some form definitions are missing. Did you run 'rails driver:hmis:seed_definitions'? #{missing_identifiers.join(', ')}" if missing_identifiers.any?
+    end
+
+    # This method builds the QA housing workflow version 1, which is a referral workflow for housing opportunities
```
This comment has me a bit confused, I think it's the "QA" part. We are working towards this being the prod AC workflow, right?
```suggestion
    # This method builds the AC housing workflow version 1, which is a referral workflow for housing opportunities
```

martha commented on drivers/hmis/lib/ce_workflows/ac_builder.rb:
```diff
@@ -0,0 +1,381 @@
+      # Initial Client Engagement => Client Engagement
+      initial_client_engagement_task.connect_to!(client_engagement_task) # TODO CONFIRM: can't bail out from this point?
+
+      # Client Engagement => Gateway => Create CE Event (Script) (or Decline)
```
It looks like Create CE Event is no longer a script task, but a triggered event on the Client Offer Outcome task. So if I'm understanding correctly, this comment is out of date

gigxz replied:
good catch, thanks

martha commented on drivers/hmis/lib/ce_workflows/ac_builder.rb:
```diff
@@ -0,0 +1,381 @@
+      template
+    end
+
+    # Workflow for admin assign referrals to non-housing projects
```
```suggestion
    # This method builds the Admin Assign workflow, which is a workflow meant to be used for direct (outgoing) referrals to non-housing projects
```

martha commented on drivers/hmis/lib/ce_workflows/ac_builder.rb:
```diff
@@ -0,0 +1,381 @@
+        name: 'Admin Assign',
+        form_definition_identifier: CE_STEP_FORMS.fetch(:initial_outgoing_referral),
+        template: template,
+        swimlane: ce_staff_swimlane, # assignment doesn't really matter since it gets sent directly?
```
That's true, assignment doesn't matter for the first step in a direct referral workflow. Maybe it could be made optional, but that might be a can of worms. Or maybe we could add a functionless-but-clarifying "Sending Project Staff" swimlane

martha replied:
No... nevermind, that's not a good idea since we wouldn't want that swimlane to be assignable on the Contacts screen.

martha reviewed (commented):
Everything looks really great, thanks for your work wrangling config 💪

gigxz committed:
Merge branch 'release-175' into gig/workflow-updates-ac

gigxz committed:
Merge branch 'release-175' into gig/workflow-updates-ac

gigxz committed:
Add update comments to address feedback

gigxz committed:
Refactor and combine util classes

gigxz commented:
I addressed the feedback (mostly adding comments) and pushed a refactor to the CE-related utility class. I combined the previous CeBuilder into a shared util class `CeBuilderUtils` that has utility functions for building workflows **and** for other tasks that may need to be run for setup. Let me know if this looks ok! And @ttoomey you are more than welcome to take a look at the file structure if you have time! I struggle with how to name these kinds of utility modules/classes

gigxz committed:
rm comment

martha reviewed (approved):
The refactor looks good to me and I like the readme you added, thanks!

ttoomey commented on drivers/hmis/lib/ce_workflows/ac/README.md:
```diff
@@ -0,0 +1,98 @@
+provider_outcome--Client Enrolled-->confirm_success
+denial_review("Denial Review"):::pink_node
+denial_review--> referral_declined
+denial_review--> provider_outcome_2
```
Observation: it would simplify the workflow if denial review returned to provider outcome. But I'm sure there's good reason for the additional complexity.

gigxz replied:
Yes that's how it worked before. This was an explicit request to have 3 separate outcome tasks, so that the response from previous attempts could be viewed.

ttoomey commented on drivers/hmis/lib/ce_workflows/shared/ce_builder_utils.rb:
```diff
@@ -0,0 +1,183 @@
+
+      Hmis::Ce::ReferralNote.where(referral: referrals).destroy_all
+      Hmis::Ce::ReferralParticipant.where(referral: referrals).destroy_all
+      referrals.destroy_all
```
this is probably fine but destroy_all may fail silently. It may be safer to use something like `find_each(&:destroy!)`

ttoomey commented on drivers/hmis/lib/form_data/allegheny/ce_referral_steps/housing_workflow_initial_review.json:
```diff
@@ -0,0 +1,109 @@
+    {
+      "text": "Referral Result",
+      "type": "CHOICE",
+      "link_id": "referral_result",
```
Note to self, while we collect this here, it looks like it's not used to create a CE event until later in the workflow

gigxz replied:
I ended up removing these referral result questions with https://github.com/greenriver/hmis-warehouse/pull/5633/commits/ba5ebd8dff3806d477ba4828b5b9486d4f149c94 per request, more context at https://github.com/open-path/Green-River/issues/8007#issuecomment-3179358804

Now the referral result is determined based on which task it was declined from

ttoomey reviewed (approved):
looks good!

gigxz committed:
Merge branch 'release-176' into gig/workflow-updates-ac

gigxz committed:
Use destroy

gigxz committed:
Remove referral result question from earlier steps

gigxz committed:
builder util update

gigxz committed:
Remove all Referral Result user input per request

gigxz committed:
Rename README
